Possible solution to Issue #43  Geometric mask 'angular' doesn't work for Odd L value

pthak006 opened the pull request:
1. GeometricMask.angular() could raise a broadcast error or produce a miscentered mask when the grid length L is odd (and in some cases when W > 2L). This PR makes the helper grid even-width and at least as wide as the requested crop so the center index is integral and the final crop always has shape (L, W).

**The main problem**: 

- For odd L, the helper x-extent has B = 2L + 1 columns (odd).

- The crop uses int(B/2 ± W/2), which floors a half-index center and can return a slice whose width ≠ W.

- With W > 2L, the helper field is narrower than the crop request, so slicing is clamped and the result also has width ≠ W.

- Final multiply self._mask *= anglemap then fails with a shape/broadcast error
- 

**Minimal problem reproduction code**:

```
import numpy as np
from sandplover.mask import GeometricMask
L, W = 51, 120   # odd L
gm = GeometricMask(np.ones((L, W)))
gm.angular(0, np.pi/3)  # before: ValueError broadcasting (L,W) with (L, <wrong>)
```

**Possible Solution Done**:

- Build a helper grid with width B = max(W, 2L), then force B even.

- Compute x = arange(-B//2, B//2), y = arange(0, W), evaluate angles, and crop a centered window of exactly W columns using integer math:

- left = B//2 - W//2

- right = left + W

- Multiply into self._mask with a guaranteed (L, W) boolean anglemap.


The overall behavior should remain same for L = even and for L = odd there can be up to a one-column difference at the far lateral edge due to removal of the half-index rounding artifact. 

Can add additional test code for the changes if needed.

pthak006 committed:
adding possible solutio to the angular mask odd length problem

amoodie commented on sandplover/mask.py:
```diff
@@ -2324,20 +2324,29 @@ def angular(self, theta1=0, theta2=np.pi):
         """
+        # width must exceed 2x length
         if (self._L / self._W) > 0.5:
             raise ValueError("Width of input array must exceed 2x length.")
```
Can we get rid of this check given your improvements? I think this was a restriction we imposed to make things work for our use cases, but this assumption of domain shape obviously doesn't generalize.

ericbarefoot replied:
I think so.

amoodie reviewed (changes requested):
I think I understand, and it makes sense. In general, some brief explanation of what the helper coords are doing could be useful. 

Also, can you create a test for this new case ([added to the end of the GeometricMask tests](https://github.com/sandpiper-toolchain/sandplover/blob/7a06ee712fb1214e8ade6a430ae7933ef41330fc/tests/mask_test.py#L1644))? A test passing with a domain of odd number should be sufficient. If you also [address the check here](https://github.com/sandpiper-toolchain/sandplover/blob/7a06ee712fb1214e8ade6a430ae7933ef41330fc/sandplover/mask.py#L2327-L2328) and we can drop it, then a test showing a passing result for a domain with non-2x width would be sufficient.

ericbarefoot reviewed (changes requested):
I agree with Andrew, and echo his request for a test.

pthak006 committed:
addressed the review comments, removed the check as requested, however it requires a slight assumption and so change in one line of the code was needed

pthak006 commented:
Thanks a lot @amoodie @ericbarefoot  for the comments. As requested by you both I:

- Added some explanation comments for the helper grid
- Removed the check the condition that was requested for. However if we do that we silently assume **rows = W** (because of `y = np.arange(self._W)[:, None]`). That assumption was safe only because the old guard forced “W big enough". Once we drop the guard, cases with **L > W** become legal—and then the math breaks.

One example which will break because of this:
```
L, W = 60, 40        # now valid once the guard is gone
arr = np.zeros((L, W))
gmsk = GeometricMask(arr)
gmsk.angular(0, np.pi/2)   # anglemask has 40 rows, mask expects 60 → broadcast error
```
- To cover for this we need to do just one change in the current code:

`y = np.arange(self._L)[:, None]           # rows follow L, not W`

- I removed an old test case which checked for bad dims previously and added a new test case that checks for odd length case.

amoodie reviewed (approved):
Looks great!! Thanks!!

One small change: remove the note at lines 2278--2281, as this will be outdated after merge.

I'll merge as soon as that's done, thanks!

pthak006 committed:
removed not required comment lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pthak006 commented:
Thanks @amoodie for the reply. I did remove the comment lines you talked about. Sorry I forgot to remove it when I did the changes.

amoodie commented:
🎉
